src/main/java/com/aws/greengrass/certificatemanager/certificate/CISShadowMonitor.java

```diff
@@ -208,31 +261,176 @@ private void subscribeToShadowTopics() throws InterruptedException {
     }
 
-    @SuppressWarnings("PMD.AvoidCatchingGenericException")
-    private synchronized void processCISShadow(int version, Map<String, Object> desiredState) {
-        if (version == lastVersion) {
-            LOGGER.atInfo().kv(VERSION, version).log("Already processed version. Skipping cert re-generation");
-            updateCISShadowReportedState(version, desiredState);
-            return;
+    private void queueCISShadowProcessing(ShadowDeltaUpdatedEvent event) {
+        shadowProcessor.queueProcessingRequest(
+                ShadowProcessingRequest.builder()
+                        .version(event.version)
+                        .desiredState(event.state)
+                        .build());
+    }
+
+    @Data
+    @Builder
+    private static class ShadowProcessingRequest {
+        private final int version;
+        private final Map<String, Object> desiredState;
+    }
+
+    /**
+     * Responsible for processing CIS shadow state.
+     *
+     * <p>The goal of this class is to react to changes in CIS connectivity
+     * and rotate registered certificates using the new host information.
+     *
+     * <p>Shadow processing by this class is idempotent, in the sense that duplicate requests
+     * to process the same shadow version will not result in unnecessary certificate rotations.
+     * Certificate rotation will only occur if this class detects changes in connectivity info
+     * from CIS.
+     *
+     * <p>If certificate rotation is successful, CIS shadow reported state
+     * will be updated to match desired state.
+     */
+    @RequiredArgsConstructor
+    class ShadowProcessor {
+
+        /**
+         * Request from {@link CISShadowMonitor} to process shadow state.
+         */
+        private final AtomicReference<ShadowProcessingRequest> shadowProcessingRequest = new AtomicReference<>();
+
+        /**
+         * Last received CIS shadow version, used to de-duplicate processing requests.
+         */
+        private final AtomicInteger lastCISShadowVersion = new AtomicInteger();
+
+        /**
+         * Host addresses obtained from CIS, used to de-duplicate processing requests.
+         */
+        private final AtomicReference<List<String>> hostAddresses = new AtomicReference<>();
+
+        private final ScheduledExecutorService ses;
+
+        /**
+         * Scheduled task for handling shadow processing requests.
+         */
+        private Future<?> shadowProcessingTask;
+        private final Object shadowProcessingTaskLock = new Object();
+
+        /**
+         * Start processing shadow requests.
+         */
+        void start() {
+            synchronized (shadowProcessingTaskLock) {
+                stop();
+                shadowProcessingTask = ses.scheduleWithFixedDelay(
+                        this::handleShadowProcessingRequest,
+                        0L,
+                        shadowProcessingDelayMs,
+                        TimeUnit.MILLISECONDS
```

Comment:
I don't think it would take much to get rid of this SES. It seems a bit silly to wake up every 5 seconds to check for an event that should be rare. Can you process shadow events on-demand in a regular executor service? I think it should be fairly straight forward

Reply:
ah yeah that's a good point, will rework this